x-pack/legacy/plugins/reporting/server/browsers/network_policy.ts

```diff
@@ -0,0 +1,18 @@
+/*
```

Comment:
I considered doing a glob-like approach to this, however it became unwieldy and hard to get configured right. In a sense you can allow/deny requests by how the start with this implementation, and even override if you have an exact match for a certain route (if its host is denied, lets say). I think it work for all the use-cases I could think of, but happy to change this.

Reply:
I don't have any concerns with this from the security perspective; however, there are certain general "rules" which are hard to express using this syntax.

For example, it's not currently possible to block all "http" traffic, except for a specific host. You'd have to hard-code each of the routes themselves, so they'd be an exact match. If we foresee this or similar rules as a common requirements, some prior-art which might be relevant is [ufw](https://help.ubuntu.com/community/UFW) which allows you to explicitly specify allow/deny rules to represent situations like this.

Reply:
@kobelb Might want another peak here. Not 1:1 with ufw, but I think it provides the flexibility you were looking for

Reply:
Thanks @joelgriffith, will review first-thing tomorrow.

Reply:
I might be representing these rules incorrectly, but the following two tests are failing:

```
  it('allows https traffic except to something forbidden', () => {
      const allow = ['https:'];
      const deny = ['https://something-forbidden/'];
      expect(allowResponse('https://kibana.com', '192.168.1.1', allow, deny)).toBe(true);
      expect(allowResponse('https://something-forbidden/', '192.168.1.1', allow, deny)).toBe(false);
      expect(allowResponse('https://something-forbidden/some-file.html', '192.168.1.1', allow, deny)).toBe(false);
    });

    it('allows http traffic only to a specific host', () => {
      const allow = ['http://something-acceptable/'];
      const deny = ['http:'];
      expect(allowResponse('http://kibana.com', '192.168.1.1', allow, deny)).toBe(false);
      expect(allowResponse('http://something-acceptable/', '192.168.1.1', allow, deny)).toBe(true);
      expect(allowResponse('http://something-acceptable/some-file.html', '192.168.1.1', allow, deny)).toBe(true);
    });
```

UFW internally [allows the rules to be ordered](https://help.ubuntu.com/community/UFW#Advanced_Example) so that as soon as a rule is matched, it stops evaluating. 

The following theoretical syntax could be used to achieve this here:

```
it('allows https traffic except to something forbidden', () => {
    const rules = [
        { result: 'deny', host: 'something-forbidden', protocol: 'https' },
        { result: 'allow', protocol: 'https' }
    ];
    expect(allowResponse('https://kibana.com', '192.168.1.1', rules)).toBe(true);
    expect(allowResponse('https://something-forbidden/', '192.168.1.1', rules)).toBe(false);
    expect(allowResponse('https://something-forbidden/some-file.html', '192.168.1.1', rules)).toBe(false);
});

it('allows http traffic only to a specific host', () => {
    const rules = [
        { result: 'allow', host: 'something-forbidden', procotol: 'http' },
        { result: 'deny', protocol: 'http' }
    ];
    expect(allowResponse('http://kibana.com', '192.168.1.1', allow, deny)).toBe(false);
    expect(allowResponse('http://something-acceptable/', '192.168.1.1', allow, deny)).toBe(true);
    expect(allowResponse('http://something-acceptable/some-file.html', '192.168.1.1', allow, deny)).toBe(true);
});
```